src/main/scala/edu/berkeley/ce/rockslicing/RockSlicer.scala

```diff
@@ -129,75 +148,79 @@ object RockSlicer {
-    val reconstructedBlocks = (processorBlocks.tail map { block =>
-      val currentBlock = Block(origin, processorBlocks.head.updateFaces(origin))
-      val updatedBlock = Block(origin, block.updateFaces(origin))
-      val sharedProcFaces = compareProcessorBlocks(currentBlock, updatedBlock)
-      if (sharedProcFaces.nonEmpty) {
-        val currentFaces = currentBlock.faces.diff(sharedProcFaces)
-        val updatedFaces = updatedBlock.faces.diff(sharedProcFaces)
-        val allFaces = currentFaces ++ updatedFaces
-        // Check for any actual shared faces between blocks - if these exists blocks
-        // should not be merged since there is a real joint seperating the blocks
-        val nonSharedFaces =
-          allFaces.foldLeft(Seq.empty[Face]) { (unique, current) =>
-            if (!unique.exists(current.isSharedWith(_))) {
-              current +: unique
-            } else {
-              unique
-            }
-          }
-        if (allFaces.diff(nonSharedFaces).isEmpty) {
-          Some(Block(origin, nonSharedFaces))
-        } else {
-          None
-        }
+                  orphanBlocks: Seq[Block]):
+                 (Seq[Block], Seq[Block]) = {
+    val blockMatches = findMates(processorBlocks)
+    val joinedBlocks = blockMatches.map{ case (paired, _) => paired }
+    val matchIndices = blockMatches.flatMap { case (_, indices) => indices}.distinct
+    val originalPairedBlocks = matchIndices.collect(processorBlocks)
+
+    val uniqueBlocks = (processorBlocks ++ orphanBlocks).foldLeft(Seq.empty[Block]) { (unique, current) =>
+      if (!unique.exists(current.approximateEquals(_))) {
+        current +: unique
       } else {
-        None
+        unique
       }
-    }).flatten
+    }
 
-    // Remove redundant faces and remove duplicates
-    val joinedBlocks = (reconstructedBlocks map { case block @ Block(center, _) =>
-      Block(center, block.nonRedundantFaces)
-    }).filter { block => block.faces.nonEmpty } // Filters blocks that actually aren't adjacent at all,
-                                                // but shared a processor joint
+    val currentOrphanBlocks = uniqueBlocks.diff(originalPairedBlocks)
+    val newProcessorBlocks = if (originalPairedBlocks.isEmpty) {
+      processorBlocks.tail
+    } else {
+      processorBlocks.diff(originalPairedBlocks)
+    }
+
+    // Check if more than one match was found. If so, merge all matches into one block
+    val mergedJoinedBlock = if (joinedBlocks.length > 1) {
+      // Note: originalPairedBlocks.head will be equal to processorBlocks.head
+      val temporaryOrigin = (originalPairedBlocks.head.centerX,
+                             originalPairedBlocks.head.centerY,
+                             originalPairedBlocks.head.centerZ)
+      val temporaryBlocks = originalPairedBlocks.tail.map { block =>
+        Block(temporaryOrigin, block.updateFaces(temporaryOrigin))
+      }
+
+      val sharedProcessorFaces = temporaryBlocks.flatMap { block =>
```

Comment:
It's hard to understand what's going on here, I think because `compareProcessorBlocks` isn't well named. `compare` generally has a specific meaning, especially when it takes two arguments: determine which block comes first under some kind of ordering.

I think what you are doing here is generating a collection of faces that are shared by the two blocks...? You should change the name of this function to reflect that, or whatever it is that's actually going on.